부산대 BE_오선재 Step1 - 카카오 로그인

5seonjae opened the pull request:
## 구현 내용 요약

- [feat] Kakao OAuth 로그인 기능 추가

  - AuthViewController에 카카오 인가 콜백 엔드포인트 구현

    - 인가 코드 → 액세스 토큰 요청 → 사용자 정보 조회 → JWT 발급 → HttpOnly 쿠키 설정 → 리다이렉트

  - KakaoTokenResponseDto, KakaoUserResponseDto DTO 정의

  - KakaoOAuthService 구현 및 예외 처리용 KakaoOAuthException 추가

  - Member 엔티티에 kakaoId, nickname 필드 추가 및 관련 생성자/Getter 반영

  - MemberRepository에 findByKakaoId 메서드 추가

  - MemberService.loginWithKakao() 로직 구현

  - 로그인 페이지(login.html)에 카카오 로그인 버튼 블록 삽입

  - 단위 테스트 작성

    - KakaoOAuthServiceTest

    - MemberServiceTest

  - build.gradle에 spring-boot-starter-webflux, okhttp3-mockwebserver 의존성 추가

---

## 테스트 및 검증

- `KakaoOAuthServiceTest`, `MemberServiceTest` 단위 테스트 구현

---

## 카카오 앱 설정

<img width="1250" height="423" alt="image" src="https://github.com/user-attachments/assets/bad54259-0770-4fa5-8e8e-753ef1d4cba7" />

<img width="1261" height="765" alt="image" src="https://github.com/user-attachments/assets/e2be8d65-f4b3-453e-adfe-6b111a32c3ab" />

<img width="1209" height="511" alt="image" src="https://github.com/user-attachments/assets/f67afb6a-e686-4a92-89ae-ad5ea375a317" />

<img width="1200" height="373" alt="image" src="https://github.com/user-attachments/assets/a0076600-71d7-488e-8426-ae2fd1f92e61" />

<img width="1196" height="667" alt="image" src="https://github.com/user-attachments/assets/a99de9f9-98d1-4ddb-9896-a7d7b060069f" />

<img width="1198" height="503" alt="image" src="https://github.com/user-attachments/assets/d7f84d08-6a78-4388-a594-5e4afe748138" />

---

## 실제 UI 화면

<img width="236" height="269" alt="image" src="https://github.com/user-attachments/assets/313f3794-3c2e-48c3-a8d1-c4f5e8d42c5b" />

<img width="656" height="781" alt="image" src="https://github.com/user-attachments/assets/7f72a59c-c8d9-4130-ac84-623a25b0f5d9" />

<img width="523" height="742" alt="image" src="https://github.com/user-attachments/assets/bf2567e3-d137-479f-8fb3-c3f75a842cb6" />

<img width="1920" height="871" alt="image" src="https://github.com/user-attachments/assets/b11ba58f-398a-4499-b844-fe9b5aa4f563" />

<img width="807" height="419" alt="image" src="https://github.com/user-attachments/assets/74063c83-9056-455d-acc5-21d6c67d094b" />

정상적으로 member 테이블에 카카오로 로그인한 회원의 데이터가 담기게 됩니다.

---

## 추가 설명

### **`WebClient` 를 선택한 이유**

- RestClient 의 종류와 특징

1. RestClient

  - RestTemplate 뒤를 잇는 동기 방식의 새로운 클라이언트
  - Java 11+ HttpClient ( JDK 기본 API ) 위에 플루언트 API를 씌운 형태
  - `RestClient.builder()` → 인터페이스로 타입 세이프 하게 호출 정의 가능
  - RestClient 는 WebClient 만큼 무겁지 않으면서도, RestTemplate 보다 모던한 API, WebClient 보다 단순한 구조
  - 장점: 동기이면서도 모던한 JDK HTTP 클라이언트 활용, 가볍고 직관적
  - 단점: 아직 생소하고, 레거시 RestTemplate 와 완벽 호환은 아님

2. WebClient

  - 비동기 ( reactive ), 논블로킹 ( non‑blocking ) 방식의 플루언트 API
  - Reactor 기반으로 만들어져, Flux/Mono 형태로 응답을 처리
  - 스케일 아웃 ( scale‑out ) 구조나 고성능 I/O 환경에 적합
  - `WebClient.builder()` 로 공통 설정 ( baseUrl, default header, 필터 등 ) 을 한 번만 정의하고 재사용
  - 장점: 반응형 파이프라인, back‑pressure, 커넥션 풀링 내장, 기능 확장성 ↑
  - 단점: 사용법이 다소 복잡, 간단한 동기 호출엔 오버헤드

3. RestTemplate

  - 동기 ( synchronous ) 방식의 템플릿 API
  - `getForObject()`, `postForEntity()` 같은 메서드 호출 위주
  - 내부에서 HTTP 요청이 끝날 때까지 스레드를 블로킹
  - Spring 3.0 부터 쓰이던 전통적 방식으로, 간단한 동기 호출에 여전히 널리 사용
  - 장점: 배우기 쉽고, 코드도 직관적
  - 단점: Reactive 흐름과 잘 어울리지 않으며, 성능 튜닝 ( 커넥션 풀 등 ) 을 직접 해 줘야 함

4. HTTP Interface

  - 인터페이스 선언만으로 REST 호출 메서드를 정의
  - 어노테이션 ( `@GetExchange`, `@PostExchange` 등 ) 을 붙이면 스프링이 런타임에 프록시 구현체를 자동 생성
  - 코드가 가장 간결해지고, HTTP 호출을 타입 세이프하게 처리
  - 장점: 서비스 계층 코드가 깔끔해지고, API 스펙 문서처럼 바로 읽을 수 있음
  - 단점: 아직 실험적 ( 6.1 기준 ), 커스터마이징이 제한적일 수 있음

---

- **WebClient 를 선택한 이유**

  - 비동기/블로킹 둘 다 OK
    - 토큰 교환·유저 조회는 I/O 가 묶이기 쉬운 작업인데, non‑blocking 으로 처리하고 필요할 땐 `.block()` 으로 동기처럼 쓸 수 있음
  - 공통 설정 재사용성
    - `.mutate()` 나 `WebClient.builder()` 를 통해 baseUrl, default header, 응답/요청 필터 등을 한 곳에 모아 두기 편함
  - 커넥션 풀, 타임아웃, 리트라이 같은 고급 설정이 내장되어 있어, 운영 환경에서 안정적으로 쓸 수 있음
  - 생태계 호환성
    - Spring Security, Spring Cloud Gateway 등 리액티브 스택과도 자연스럽게 맞물림

성능·안정성 ( 논블로킹 I/O, 커넥션 관리 ), 유지보수성 ( 재사용 설정, 필터 AOP ), 확장성 ( 전면 리액티브 전환이나 다중 엔드포인트 연동 ) 측면에서 **`WebClient`** 가 가장 적절하다고 생각했습니다.

---

리뷰 부탁드립니다!

5seonjae committed:
docs(readme): .env 설정 안내 및 카카오 로그인 가이드 추가

- 프로젝트 루트에 `.env` 파일 생성 및 필수 환경 변수(JWT_SECRET, JWT_EXPIRATION_MS, KAKAO_REST_API_CLIENT_ID) 설정 방법 안내
- Step1 카카오 로그인 기능 요구사항 체크리스트 추가
- HTTP API 엔드포인트 및 인가 URL 예시 문서화

5seonjae committed:
feat(kakao): 카카오 OAuth 로그인 기능 추가

- AuthViewController에 콜백 엔드포인트 구현 (인가 코드 → JWT 발급 → HttpOnly 쿠키 설정 및 리다이렉트)
- KakaoTokenResponseDto, KakaoUserResponseDto DTO 정의
- KakaoOAuthService 구현 및 KakaoOAuthException 추가
- Member 엔티티에 kakaoId, nickname 필드 및 생성자/Getter 추가
- MemberRepository에 findByKakaoId 메서드 추가
- MemberService에 loginWithKakao 로직 구현 (회원 조회/가입 후 JWT 발급)
- login.html에 카카오 로그인 버튼 링크 블록 추가
- KakaoOAuthServiceTest, MemberServiceTest 단위 테스트 작성
- build.gradle에 spring-boot-starter-webflux, okhttp3-mockwebserver 의존성 추가

5seonjae committed:
docs(readme): README 환경설정 및 카카오 로그인 가이드 업데이트

- `.env` 환경 변수 테이블 수정 및 필수 키 설명 보강
- 기능 요구사항 체크리스트 항목 수정 및 상태 반영
- HTTP API 섹션 전면 재구성(엔드포인트, 설명, 예시 보강)
- 인가 URL 예시 가독성 개선 및 쿼리 파라미터 설명 추가

Gyuchool commented on src/main/java/gift/controller/view/AuthViewController.java:
```diff
@@ -86,4 +88,34 @@ public String logout(HttpServletResponse httpServletResponse) {
+            String jwt = memberService.loginWithKakao(code);
+
+            // 2) HttpOnly 쿠키에 JWT 담기
+            Cookie authCookie = new Cookie("AUTH", jwt);
```
ResponseCookieBuilder를 사용하는 것과 그냥 new Cookie를 사용하는 방식에는 어떤 차이가 있나요?🙄

5seonjae replied:
리뷰 감사합니다! 😊

`ResponseCookieBuilder` ( = `ResponseCookie.from(…)` ) 방식과 `new Cookie(…)` ( Servlet API ) 방식의 주요 차이점을 정리해보겠습니다.

1. API·객체 모델

| 구분 | new Cookie(...) (Servlet API) | ResponseCookieBuilder (Spring) |
| -- | -- | -- |
| 클래스 | `javax.servlet.http.Cookie` | `org.springframework.http.ResponseCookie` |
| 객체 상태 | mutable ( setPath, setSecure 등 세터로 변경 ) | immutable ( 빌드 후 상태 변경 불가 ) |
| 헤더 등록 | `response.addCookie(cookie)` → 컨테이너가 Set‑Cookie 헤더 생성 | `response.addHeader(HttpHeaders.SET_COOKIE, cookie.toString())` ( MVC ) 또는 `ServerHttpResponse#addCookie(cookie)` (WebFlux) |

2. 지원 속성 비교

| 속성 | Servlet Cookie | ResponseCookie (Builder)
| -- | -- | --
| HttpOnly | ✅ `setHttpOnly(true)` | ✅ `.httpOnly(true)`
| Secure | ✅ `setSecure(true)` | ✅ `.secure(true)`
| Path | ✅ `setPath(...)` | ✅ `.path(...)`
| Domain | ✅ `setDomain(...)` | ✅ `.domain(...)`
| Max‑Age | ✅ `setMaxAge(seconds)` | ✅ `.maxAge(Duration)`
| Expires | ❌ ( 미지원 ) | ✅ 자동 계산
| SameSite | ❌ ( 미지원 ) | ✅ `.sameSite("Lax")` 등
| Extensions | ❌ | ✅ `.attr("X-Foo","bar")`

3. 장·단점

🟢 `new Cookie(...)`

- 장점

  - 서블릿 컨테이너 ( Standard Servlet API ) 에 의존 → 어디서나 동일하게 동작함

  - 코드가 짧고 간단함 ( HttpOnly/Secure 만 설정할 경우 )

  - `response.addCookie()` 로 자동 포맷팅

- 단점

  - `SameSite`, `Expires` 등 최신 쿠키 속성 미지원

  - mutable 객체 → 잘못 공유 시 스레드 안전성 이슈

  - 빌더 API 부재로 코드 가독성·테스트 불편

🟢 `ResponseCookieBuilder`

- 장점

  - 불변 ( immutable ) · 플루언트 빌더 → 가독성↑, 테스트 편리

  - `SameSite`, `Expires`, `Priority` 등 최신 RFC 속성 기본 지원

  - WebFlux ( reactive ) 환경에 네이티브 최적화

  - 확장 속성( `.attr(...)` ) 추가가 용이함

- 단점

  - Spring Framework API 의존 → 순수 Servlet 프로젝트에는 불필요

  - `addHeader()` 직접 호출 → 컨테이너 자동 관리 ( 중복 처리 등 ) 를 못씀

  - 설정할 속성이 적으면 코드가 다소 장황해짐

4. 언제 무엇을 써야 할까?

단순히 HttpOnly+Secure 정도만 필요할 때에는 `new Cookie(...)` 방식을 사용합니다.

`SameSite` 설정이 필요하거나 `Expires` 헤더가 필요할 때, `Spring WebFlux(reactive)` 환경, 테스트 코드에서 쿠키 속성 조합을 깔끔히 검증할 때에는 `ResponseCookieBuilder` 를 사용합니다.

---

### `kakaoCallback` 메서드 구현

`kakaoCallback` 의 현재 구현은 다음과 같이 `javax.servlet.http.Cookie` 를 직접 생성하고 있습니다:

```java
Cookie authCookie = new Cookie("AUTH", jwt);
            authCookie.setHttpOnly(true);
            authCookie.setPath("/");
            response.addCookie(authCookie);
```

하지만 운영 환경에서 보안·호환성·유지보수를 고려하면, 저는 `ResponseCookieBuilder` 방식을 사용하는 것이 더 적절하다고 생각합니다. 그 이유를 정리하면 다음과 같습니다:

1. SameSite 설정 지원

- 크로스사이트 컨텍스트에서 쿠키의 전송 여부를 제어하는 SameSite 속성을 명시적으로 설정할 수 있습니다.

- `new Cookie()` API 에는 이 옵션이 없어, 헤더를 수동으로 조작해야 하는 번거로움과 실수 위험이 있습니다.

2. 불변(Immutable)·빌더 패턴

- `ResponseCookie` 는 불변 객체이며, 빌더를 통해 일관된 속성 설정이 가능합니다.

- 체인형 메서드 호출로 설정 누락을 방지하고, 여러 스레드에서 안전하게 사용할 수 있습니다.

3. Servlet/Reactive 양쪽 호환

- `ResponseCookie` 는 `Servlet ( HttpServletResponse )` 뿐만 아니라 `WebFlux ( ServerHttpResponse )` 환경에서도 동일한 방식으로 사용할 수 있어, 스택 전환 시 유지보수가 용이합니다.

4. RFC 6265 규격 준수

- `ResponseCookie.toString()` 은 표준 Set‑Cookie 헤더 포맷을 자동 생성해 주므로, 문자열 포맷 실수를 줄여 줍니다.

이러한 점들을 고려하여 다음과 같은 방식으로 수정하려고 합니다.

```java
ResponseCookie cookie = ResponseCookie.from("AUTH", jwt)
    .httpOnly(true)
    .secure(true)
    .path("/")
    .maxAge(Duration.ofHours(1))
    .sameSite("Lax")
    .build();
response.addHeader(HttpHeaders.SET_COOKIE, cookie.toString());
```

---

Gyuchool replied:
정말 열심히 학습하셨군요!!💯💯

Gyuchool commented on README.md:
```diff
@@ -1 +1,137 @@
\ No newline at end of file
+# spring-gift-order
+
+### .env 파일 준비
```
정리를 잘하셨군요 👍

Gyuchool commented on src/main/java/gift/controller/view/AuthViewController.java:
```diff
@@ -86,4 +88,34 @@ public String logout(HttpServletResponse httpServletResponse) {
     }
+
+    /**
+     * 카카오 OAuth 콜백 처리
```
주석은 제거하시는게 좋아보이는데 주석이 꼭 필요한걸까요?🙄
선재님은 주석을 언제 사용하시나요?

5seonjae replied:
리뷰 감사합니다! 😊
해당 주석은 개발 중 로직 확인을 위해 임시로 추가했던 것이라, 기능 구현을 완료하고 깜빡하고 남겨두었네요. 
불필요한 주석이 코드 가독성을 오히려 해칠 수 있고, 메서드·변수 이름만으로 충분히 의도를 드러낼 수 있다는 점에 전적으로 공감합니다.
말씀해주신 대로 이번 기회에 프로젝트 전반의 불필요한 주석을 모두 제거하고, 정말 필요한 경우에만 간결하고 명확한 주석을 남기도록 하겠습니다! 🚀

Gyuchool commented on src/main/java/gift/controller/view/AuthViewController.java:
```diff
@@ -86,4 +88,34 @@ public String logout(HttpServletResponse httpServletResponse) {
+
+            // 3) 로그인 후 메인 페이지(또는 원하는 URL)로
+            response.sendRedirect("/products");
+        } else {
```
객체지향 생활 체조 원칙!
> 규칙 2: else 예약어를 쓰지 않는다.

else를 제거해볼까요?

5seonjae replied:
리뷰 감사합니다! 🙌 
제가 객체지향 생활 체조 원칙의 중요한 점을 놓친 것 같습니다.
else 블록을 제거하고 가드 클로즈(guard clause) 형태로 바꿔 보았습니다. 
중첩을 줄여서 흐름이 좀 더 직관적으로 읽힐 것 같습니다.

```java
@GetMapping("/")
public void kakaoCallback(
    @RequestParam(value = "code", required = false) String code,
    HttpServletResponse response
) throws IOException {
    // 1) code가 없으면 바로 로그인 페이지로
    if (code == null) {
        response.sendRedirect("/login");
        return;
    }

    // 2) code가 있을 때만 JWT 발급 → HttpOnly 쿠키 저장 → 메인 페이지로
    String jwt = memberService.loginWithKakao(code);
    ResponseCookie cookie = ResponseCookie.from("AUTH", jwt)
        .httpOnly(true)
        .secure(true)
        .path("/")
        .maxAge(Duration.ofHours(1))
        .sameSite("Lax")
        .build();
    response.addHeader(HttpHeaders.SET_COOKIE, cookie.toString());
    response.sendRedirect("/products");
}
```

이런 형태로 수정해서 커밋하겠습니다! 🚀

Gyuchool commented on src/main/java/gift/controller/view/AuthViewController.java:
```diff
@@ -86,4 +88,34 @@ public String logout(HttpServletResponse httpServletResponse) {
+
+            // 2) HttpOnly 쿠키에 JWT 담기
+            Cookie authCookie = new Cookie("AUTH", jwt);
+            authCookie.setHttpOnly(true);
```
쿠키 설정을 지금처럼 설정한 이유가 있을까요?

httpOnly, secure, sameSite 설정은 각각 무엇이 다를까요?🤔

5seonjae replied:
리뷰 감사합니다! 😊
아래에 이전 코드와 지금 수정된 코드를 함께 설명드리면서, 각 속성을 왜 설정했는지 정리해보겠습니다.

### 1. 이전 코드 ( Servlet API `Cookie` )

```java
Cookie authCookie = new Cookie("AUTH", jwt);
authCookie.setHttpOnly(true);
authCookie.setPath("/");
response.addCookie(authCookie);
```

- HttpOnly

  - `setHttpOnly(true)`
    → 자바스크립트에서 document.cookie 로 JWT를 읽지 못하게 막아 XSS 공격 시 탈취를 방지

- Path

  - `setPath("/")`
    → 애플리케이션의 모든 URL 경로에서 쿠키가 자동 전송되도록 최상위 경로로 지정

- Secure, SameSite 미설정

  - 로컬 개발 편의상 HTTPS 강제(setSecure(true))와 SameSite 속성을 나중에 보완하려고 일단 생략

  - Servlet Cookie API 자체가 SameSite 를 지원하지 않아, CSRF 대응은 별도의 CSRF 토큰 검증으로 보완 예정

최소한의 XSS 방어 ( HttpOnly ) 와 경로 설정만 적용해 로컬 개발과 기본 동작을 우선 확보하기 위해 이렇게 구현했었습니다.

---

### 2. 수정된 코드 ( Spring `ResponseCookieBuilder` )

```java
ResponseCookie cookie = ResponseCookie.from("AUTH", jwt)
    .httpOnly(true)                   
    .secure(true)                     
    .path("/")                        
    .maxAge(Duration.ofHours(1))      
    .sameSite("Lax")                  
    .build();
response.addHeader(HttpHeaders.SET_COOKIE, cookie.toString());
```

- HttpOnly
→ 이전 동일: JS 접근 차단으로 XSS 방어

- Secure
→ HTTPS 환경에서만 전송하도록 강제, MITM ( 중간자 ) 공격으로부터 보호

- Path
→ "/" 로 지정해 전체 애플리케이션에서 쿠키 전송

- Max‑Age
→ 1시간 ( `Duration.ofHours(1)` ) 뒤 만료시켜 세션 지속시간 제어

- SameSite=Lax
→ 링크 클릭 같은 “정상 내비게이션”에는 쿠키 전송, `<img>`·`fetch` 같은 일부 크로스사이트 요청에는 차단해 CSRF 공격 위험 감소

`ResponseCookie` 빌더로 HttpOnly, Secure, Path, Max‑Age, SameSite 까지 한 번에 설정해 XSS, MITM, CSRF 등 주요 공격 벡터를 두루 방어하면서도, 세션 만료 정책과 동일 사이트 정책을 명시적으로 다룰 수 있게 개선했습니다.

---

### httpOnly, secure, sameSite

| 속성 | 역할 | 보안 효과 |
| -- | -- | -- |
| HttpOnly | JS ( document.cookie ) 접근 차단 | XSS 공격 시 쿠키 탈취 방지 |
| Secure | HTTPS 연결에서만 전송 | 네트워크 스니핑 ( MITM ) 방지 |
| SameSite | 크로스사이트 전송 제어 | CSRF 공격 위험 감소 – `Strict` : 완전 차단 – `Lax` : 링크 클릭만 허용 – `None` : 모든 요청에 전송 ( 단, Secure 필수 ) |

Gyuchool replied:
정리가 깔끔하게 해서 더 할말이 없네요..💯💯
SameSite 설정은 브라우저마다 기본값과 동작 방식이 다를 수 있어서 적용할 때 확인이 필요하기도 합니다!

Gyuchool commented on src/main/java/gift/dto/api/KakaoTokenResponseDto.java:
```diff
@@ -0,0 +1,11 @@
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public record KakaoTokenResponseDto (
```
`@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)`
해당 어노테이션을 학습후 적용해봐도 좋겠네요!

5seonjae replied:
리뷰 감사드립니다! 🙏

지금은 각 필드마다 `@JsonProperty` 로 snake_case 매핑을 해두었는데,
클래스 레벨에 `@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)` 를 붙이면 아래와 같이 훨씬 간결하게 리팩토링할 수 있겠네요:

```java
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
public record KakaoTokenResponseDto(
    String accessToken,
    String refreshToken,
    long expiresIn,
    String tokenType
) {}
```

학습 후 실제 코드에 적용해서 불필요한 어노테이션을 제거해보겠습니다! 🚀

Gyuchool commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -138,4 +139,12 @@ public ResponseEntity<Map<String, String>> handleOptionProductMismatch(OptionPro
+        ProblemDetail pd = ProblemDetail.forStatus(HttpStatus.BAD_GATEWAY);
+        pd.setTitle("카카오 OAuth 처리 오류");
+        pd.setDetail(ex.getMessage());
+        return ResponseEntity.status(HttpStatus.BAD_GATEWAY).body(pd);
```
502 에러를 내려주는게 적합하다고 생각하신 이유가 궁금합니다!🙄
500과 502 각각 언제 내려주는게 적합하다고 생각하시나요

5seonjae replied:
리뷰 감사합니다! 🙏

`502 Bad Gateway` 를 선택한 이유는, 우리 애플리케이션이 게이트웨이 역할로 카카오 OAuth 서버를 호출하다가 업스트림 ( 카카오 ) 에서 5xx 에러를 반환받았을 때, 그 사실을 클라이언트에게 명확히 전달하기 위함입니다. 반면에 `500 Internal Server Error` 는 “우리 서버 자체에서 예기치 못한 버그나 처리 로직 오류”가 발생했을 때 내려주는 것이 RESTful 관점에서 적절합니다.

- 클라이언트가 유효하지 않은 code 를 보냈을 때  ( 400 Bad Request )
- 엑세스 토큰 유효성 검사 실패 시 ( 401 Unauthorized )
- 카카오 API 호출 중 5xx 응답 ( 서버 에러 ) 을 받았을 때 ( 502 Bad Gateway )
- 카카오 서비스 자체가 다운돼 응답이 없을 때 ( 503 Service Unavailable )
- 카카오 호출이 타임아웃될 때 ( 504 Gateway Timeout )
- 애플리케이션 내부에서 예기치 않은 에러가 났을 때 ( 500 Internal Server Error )

이렇게 구분해 두면 `502 Bad Gateway` 의 경우에는 “카카오 쪽 문제” 고, `500 Internal Server Error` 의 경우에는 “우리 쪽 문제” 를 HTTP 상태코드만 보고도 빠르게 파악할 수 있어, 운영·디버깅이 훨씬 편해집니다. 🚀

---

이 점을 참고해서 예외 처리에 관련된 코드를 수정하려 합니다.

```java
// KakaoOAuthService.java
public KakaoTokenResponseDto exchangeCodeForToken(String code) {
        return webClient.post()
            .uri(authUrl + "/oauth/token")
            .contentType(MediaType.APPLICATION_FORM_URLENCODED)
            .body(BodyInserters
                .fromFormData("grant_type", "authorization_code")
                .with("client_id", clientId)
                .with("redirect_uri", redirectUri)
                .with("code", code))
            .retrieve()
            .onStatus(HttpStatusCode::isError, resp ->
                resp.bodyToMono(String.class)
                    .flatMap(body -> Mono.error(new KakaoOAuthException(
                        resp.statusCode(),
                        "토큰 교환 실패: " + resp.statusCode() + " - " + body
                    )))
            )
            .bodyToMono(KakaoTokenResponseDto.class)
            .block();
    }

    public KakaoUserResponseDto fetchUserInfo(String accessToken) {
        return webClient.get()
            .uri(apiUrl + "/v2/user/me")
            .header("Authorization", "Bearer " + accessToken)
            .retrieve()
            .onStatus(HttpStatusCode::isError, resp ->
                resp.bodyToMono(String.class)
                    .flatMap(body -> Mono.error(new KakaoOAuthException(
                        resp.statusCode(),
                        "유저 정보 조회 실패: " + resp.statusCode() + " - " + body
                    )))
            )
            .bodyToMono(KakaoUserResponseDto.class)
            .block();
    }
```

```java
public class KakaoOAuthException extends RuntimeException {
    private final HttpStatusCode statusCode;

    public KakaoOAuthException(HttpStatusCode statusCode, String message) {
        super(message);
        this.statusCode = statusCode;
    }

    public HttpStatusCode getStatusCode() {
        return statusCode;
    }
}
```

```java
// GlobalExceptionHandler.java
@ExceptionHandler(KakaoOAuthException.class)
    public ResponseEntity<ProblemDetail> handleKakaoOAuth(KakaoOAuthException ex) {
        HttpStatusCode upstream = ex.getStatusCode();
        HttpStatus status = mapToStatus(upstream);

        ProblemDetail pd = ProblemDetail.forStatus(status);
        pd.setTitle("카카오 OAuth 처리 오류");
        pd.setDetail(ex.getMessage());
        return ResponseEntity.status(status).body(pd);
    }

    @ExceptionHandler(ResourceAccessException.class)
    public ResponseEntity<ProblemDetail> handleTimeout(ResourceAccessException ex) {
        if (ex.getCause() instanceof SocketTimeoutException) {
            ProblemDetail pd = ProblemDetail.forStatus(HttpStatus.GATEWAY_TIMEOUT);
            pd.setTitle("카카오 OAuth 타임아웃");
            pd.setDetail("카카오 API 호출이 시간 내에 응답하지 못했습니다.");
            return ResponseEntity.status(HttpStatus.GATEWAY_TIMEOUT).body(pd);
        }
        return handleInternal(ex);
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ProblemDetail> handleInternal(Exception ex) {
        ProblemDetail pd = ProblemDetail.forStatus(HttpStatus.INTERNAL_SERVER_ERROR);
        pd.setTitle("서버 내부 오류");
        pd.setDetail(ex.getMessage());
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(pd);
    }

    private HttpStatus mapToStatus(HttpStatusCode upstream) {
        if (upstream.is4xxClientError()) {
            return (upstream.value() == HttpStatus.UNAUTHORIZED.value()
                ? HttpStatus.UNAUTHORIZED
                : HttpStatus.BAD_REQUEST);
        }
        if (upstream.is5xxServerError()) {
            return HttpStatus.BAD_GATEWAY;
        }
        return HttpStatus.INTERNAL_SERVER_ERROR;
    }
```

### 코드 설명

- `KakaoOAuthException` / `mapToStatus`

  - `KakaoOAuthException` 에는 4xx, 5xx 번대의 예외들이 들어오게 됩니다. 

  - 이 예외들 중에서 `401 UNAUTHORIZED` 는 따로 처리하고, 나머지 4xx 번대의 예외들은 `400 BAD_REQUEST` 로 묶어서 처리합니다.

  - 카카오 API가 5xx 에러를 응답하는 건 “우리 서버가 게이트웨이 역할을 하다 업스트림 문제를 받았다”는 의미이기 때문에, 5xx 번대 예외들을 502 BAD_GATEWAY 로 묶어서 처리합니다.

  - 혹시 다른 경로로 잘못 전달됐다면, 최종 안전망으로 `500 Internal Server Error` 를 내려 “우리 쪽에서 처리할 수 없는 예외” 임을 표시합니다.

- `handleTimeout` / `handleInternal`

  - 네트워크·타임아웃 오류 구분

    - `ResourceAccessException` 는 `RestTemplate/WebClient` 가 원격 서버 ( 여기서는 카카오 ) 에 요청을 보내는 도중에 커넥션 자체가 실패했거나, 응답을 제때 받지 못해 타임아웃이 발생했을 때 던져집니다.

    - 이 중 `SocketTimeoutException` 인 경우만 골라서 `504 Gateway Timeout` ( `HttpStatus.GATEWAY_TIMEOUT` ) 을 내려 줌으로써 “카카오 호출이 응답 지연으로 실패했다”는 점을 명확히 알려 줍니다.

    - 만약 이걸 그냥 `500 Internal Server Error` 로 묶어 버리면, 클라이언트는 네트워크 문제인지 내부 버그인지 구분할 수 없습니다.

  - 나머지 예외 ( 내부 버그 ) 통합 처리

    - `Exception` 핸들러 ( `handleInternal` ) 는 `KakaoOAuthException` 이나 `ResourceAccessException` 에 걸리지 않는 모든 예외 ( NPE, 파싱 오류 등 ) 를 잡습니다.

    - 이 경우 `500 Internal Server Error` ( `HttpStatus.INTERNAL_SERVER_ERROR` ) 를 내려서 “우리 애플리케이션 내부에서 예기치 못한 오류가 발생했다”는 점만 일관되게 전달합니다.

    - 이렇게 하면 불필요하게 스택트레이스나 세부 구현이 노출되지 않고, API 소비자 입장에서는 항상 JSON ProblemDetail 폼으로 일관된 에러 응답을 받게 되어 편리합니다.

---

### 추가 설명

- “클라이언트 요청이 잘못된 것” 전반을 400으로 묶어서 단순화하는 이유

  - 불필요한 복잡성

    - 예를 들어 카카오가 400, 403, 404, 429 등 여러 4xx 코드를 쏘면, 우리 클라이언트 ( 웹·앱 ) 는 “이게 뭘 의미하는 거지?” 하고 일일이 케이스를 나눠 처리해야 합니다.

  - 업스트림 내부 사양 노출

    - `404 Not Found` 는 “해당 리소스 없음” 을 의미하는데, 카카오 API 에서는 이 의미가 그대로 우리 고객에게도 유효하지 않을 수 있습니다. ( 예 : 인가 코드로 리소스 접근을 시도했을 때 )

    - 이렇게 되면 API 설계 시 “카카오 내부 동작”이 노출되어, 우리 서비스의 일관된 계약 ( Contract ) 이 깨지게 됩니다.

  - 유지보수 난이도 증가

    - upstream 변경에 따라 더 많은 4xx 를 케이스별로 추가해야 하고, 코드가 급격히 복잡해집니다.

Gyuchool replied:
저도 동감합니다😊 정리 깔끔하네요..!

Gyuchool commented on src/main/java/gift/service/KakaoOAuthService.java:
```diff
@@ -0,0 +1,81 @@
+@Service
+public class KakaoOAuthService {
+
+    @Value("${rest-api.clientId}")
```
해당 프로퍼티만 rest-api. prefix를 둔 이유가 있으실까요?

5seonjae replied:
리뷰 감사합니다! 😊

사실 카카오 로그인용 키가 총 네 종류 ( REST API 키, JavaScript 키, Admin 키, 네이티브 앱 키 )  있었는데, 그중 REST API 키만 따로 분리해 두려고 `rest-api.clientId` 라는 접두어를 썼습니다.

하지만 아래와 같이 다른 카카오 관련 프로퍼티들이 모두 `kakao.*`로 묶여 있는 걸 보니, REST API 키도 `kakao.client-id` 로 통일하는 것이 가독성과 설정 일관성 측면에서 더 좋을 것 같습니다.

이렇게 하면 “카카오 OAuth 에 필요한 모든 설정이 `kakao.*`” 로 한눈에 모여서, 유지보수할 때 더 직관적일 것 같습니다. 이 방향으로 리팩토링해 보겠습니다! 🚀

Gyuchool replied:
좋은 것 같아요! 다른 플랫폼의 oauth를 추가할때도 구별되어서 확장성에 좋을 것 같아요!

Gyuchool commented on src/main/java/gift/service/KakaoOAuthService.java:
```diff
@@ -0,0 +1,81 @@
+            .retrieve()
+            .onStatus(HttpStatusCode::isError, resp ->
+                resp.bodyToMono(String.class)
+                    .flatMap(body -> Mono.error(
```
webClient를 사용하셨군요..! 
webClient의 각 메서드가 어느 역할을 하는지도 학습하셨을까요?
retrieve, flatMap, Mono, etc..

webClient를 사용해보면서 느끼신 점이 궁금합니다:)

5seonjae replied:
WebClient 를 쓰면서 제가 느끼고 공부한 주요 포인트를 정리해 보았습니다.

### 1. WebClient 주요 메서드 역할

- `WebClient.builder()` / `WebClient.create()`

  - 인스턴스 생성과 공통 설정 ( baseUrl, defaultHeader 등 )

- HTTP 메서드 선택 ( `.get()`, `.post()`, … )

  - 어떤 HTTP 메서드를 사용할지 지정

- `.uri(...)`

  - URI 템플릿과 변수 바인딩

- 요청 설정 ( `.headers()`, `.body(…)` )

  - 헤더, 쿼리파라미터, 폼 데이터, JSON 바디 등

- `.retrieve()`

  - 실제 호출을 트리거

  - 상태 코드 ( 2xx/4xx/5xx ) 에 따른 후속 처리를 쉽게 연결

- `.onStatus(predicate, handler)`

  - predicate ( 예 : `HttpStatusCode::isError` ) 에 걸리는 상태코드를 잡아 클라이언트 응답 ( `ClientResponse` ) 을 기반으로 `Mono.error(...)` 를 만들어 줌

- `.bodyToMono(Class<T>)` / `.bodyToFlux(Class<T>)`

  - 응답 바디를 `Mono<T>` ( 단건 ) 또는 `Flux<T>` ( 스트림 ) 로 변환

- Mono / Flux 연산자

  - `map()` : 동기적 데이터 변환 ( T → R )

  - `flatMap()` : 비동기적 데이터 변환 ( T → Mono<R> ) 후 결과 풀어주기

  - 그 외 filter(), zip(), onErrorResume() 등 풍부한 조합 가능

- `.block()` / `.subscribe()`

  - `block()` : 테스트나 간단 스크립트에서 동기 호출할 때 사용 ( 프로덕션 코드에선 가급적 지양 )

  - `subscribe()` : 완전 비동기 방식으로 결과를 소비할 때 사용

---

### 2. Reactor 타입 이해

`Mono<T>` : 0 또는 1개의 데이터를 비동기 발행

`Flux<T>` : 0개 이상의 데이터를 순차적으로 비동기 발행

장점

- 논블로킹 I/O 기반으로, 스레드 효율 극대화

- 에러, 완료, 취소 신호까지 포함한 일관된 이벤트 흐름

주의할 점

- 학습 곡선이 있고, 잘못 쓰면 `block()` 남용으로 블로킹 로직이 될 수 있음

---

### 3. 느낀 점

가독성 & 유지보수성 측면에서 체인 형태로 “요청 → 상태분기 → 바디 추출 → 변환” 흐름이 한눈에 보이기 때문에 예외 처리 ( `onStatus` ) 와 비즈니스 로직 ( `flatMap` ) 분리가 명확해진 것을 느꼈습니다.

`onStatus` 로 HTTP 오류, `onErrorResume` 으로 런타임 예외까지 세밀하게 제어 가능하다는 것을 통해 에러 처리의 유연함을 알게 되었습니다.

비동기·논블로킹 측면에서 스레드를 차지하지 않고 대규모 동시 요청을 처리할 수 있어, 고성능 API 클라이언트에 적합하다는 것을 느꼈습니다.

Gyuchool replied:
그렇다면 항상 webflux를 쓰는게 좋은걸까요? 
논블로킹 I/O기반으로 처리량이 빠르겠지만 결국 `.block();` 으로 인해 블로킹 I/O처럼 사용하지 않나요?🙄

5seonjae replied:
> 그렇다면 항상 webflux를 쓰는게 좋은걸까요?

결론적으로 “무조건 WebFlux → 더 빠르다” 라는 공식은 없습니다.
논블로킹 I/O 가 큰 힘을 발휘하려면 체인의 끝 ( 컨트롤러 ) 까지 Mono/Flux 로 유지돼야 하고, JDBC 같이 블로킹 자원이 사라져야 합니다.
`block()` 을 호출해 버리면 그 지점부터는 기존 MVC + RestTemplate 패턴과 거의 동일한 블로킹 동작으로 바뀌게 됩니다.

### WebFlux ( 논블로킹 ) 가 이득이 되는 경우

1. 컨트롤러도 Mono/Flux 를 리턴하고  R2DBC·Reactive Redis 를 사용할 때
     thread‑per‑request 병목이 사라짐 → 소수 스레드로 수천 커넥션 처리

2. SSE/WebSocket·스트리밍 API 다수
    back‑pressure·keep‑alive 에 최적

3. 네트워크 레이어 제어가 복잡할 때 ( 타임아웃, Retry, CircuitBreaker )
    연산자 체인으로 선언적으로 조합 가능

---

### `.block()` 을 사용한 이유

1. 현재 Spring Data JPA ( = JDBC 블로킹 ) 를 사용하고 있기 때문에 어차피 요청‑스레드가 DB I/O 에서 block 이 되어 외부 API 를 논블로킹으로 바꿔도 전체 처리량은 크게 바뀌지 않습니다.

2. 현재 프로젝트에서는 카카오 OAuth API 를 1 회만 호출하고 있기 때문에 논블로킹 체인 ( setup 비용 ) 이 얻는 성능 효과보다 크다고 생각합니다.

즉 “로그인 한 번 호출하고 DB INSERT” 패턴에서는 단순하고 안정적인 것이 가장 큰 가치라고 생각해서 `.block()` 으로 동기 플로우를 유지하려고 했습니다.

이후에 API 확장성을 고려해서 나중에 진짜 논블로킹이 필요하면 `.block()` 만 제거하고 Mono/Flux 체인으로 전환이 가능합니다.

그리고 RestTemplate 대신 WebClient 를 쓰면 HTTP/2, 필터 DSL, Circuit‑Breaker 연동 등을 그대로 활용할 수 있습니다.

---

### ‘풀 Reactive’로 바꾸는 시기

- 외부 API 호출이 여러 개로 늘어 동시 I/O 병목이 보일 때

- JDBC 대신 R2DBC 도입, 전체 체인이 논블로킹으로 바뀔 때

- SSE·WebSocket 등 스트리밍 기능을 본격 도입할 때

---

### 정리

현재 프로젝트 규모·스택 ( JPA + MVC ) 에서는 `.block()` 으로 동기 처리하는 편이 가장 단순하고 안정적이라고 생각했습니다.

WebClient 를 사용해둔 덕분에, “외부 API가 늘어나 논블로킹 전환이 필요”해지는 순간에도 큰 구조 변경 없이 Mono/Flux 체인으로 점진적 마이그레이션이 가능합니다.

Gyuchool replied:
Tomcat 기반의 서버에서 Netty 기반 서버로 전환된다면, 단순히 외부 호출 방식뿐만 아니라 내부 구조나 처리 방식 등 여러 부분에서 변화가 클 것 같습니다..! 또한, 주문 서버의 경우, WebFlux 기반으로 전환될 가능성은 그리 높지 않다고 생각됩니다:)

개인적으로는 MVC 기반 서버에서는 RestTemplate이나 RestClient를 활용하고, WebFlux 기반 서버에서는 WebClient를 활용하는 등, 각 기술의 특성과 장점을 살리는 방식이 가장 적합하다고 생각되어 질문을 드렸습니다😊

Gyuchool commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -50,4 +56,23 @@ public boolean isAdmin(String email) {
+    @Transactional
+    public String loginWithKakao(String code) {
+        // 1) 코드 → 토큰, 토큰 → 유저 정보
+        var tokenDto = kakaoOAuthService.exchangeCodeForToken(code);
```
여기는 타입 추론을 사용하셨네요? 일관성있게 타입 명시해주는게 더 좋아보입니다!

5seonjae replied:
피드백 감사합니다! 😊

var를 써서 간결하게 표현할 생각이었는데, 서비스 레이어에서 반환 타입이 한눈에 드러나지 않으면 오히려 가독성을 해칠 수 있다는 점에 공감합니다. 
특히 DTO 명이 도메인 의미를 담고 있어 코드 이해에 도움이 되므로, 명시적으로 적어 주는 편이 더 낫겠네요.

```java
// before
var tokenDto = kakaoAuthService.exchangeCodeForToken(code);

// after
KakaoTokenResponseDto tokenDto = kakaoAuthService.exchangeCodeForToken(code);
```

이렇게 타입을 명시하도록 수정하겠습니다! 🚀

Gyuchool commented on src/test/java/gift/KakaoOAuthServiceTest.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    @Test
+    void exchangeCodeForToken_success() {
```
주석보단 `@DisplayName`으로 어떤 테스트인지 명확하게 표시해주는게 좋을 것 같습니다!

5seonjae replied:
이전 테스트들엔 항상 `@DisplayName` 을 달아 왔는데 이번 커밋에서는 깜빡한 것 같습니다;;

말씀해주신 대로 주석 대신 `@DisplayName` 을 추가해 일관성을 맞추겠습니다.

Gyuchool commented on src/test/java/gift/MemberServiceTest.java:
```diff
@@ -0,0 +1,102 @@
+import org.springframework.transaction.annotation.Transactional;
+
+@SpringBootTest
+@AutoConfigureMockMvc
```
해당 어노테이션은 어떤 역할일까요? 제거하면 어떻게 되나요?

5seonjae replied:
통합‑테스트를 위주로 작성하다 보니 두 애너테이션을 습관적으로 붙였습니다.

`@SpringBootTest` + `@AutoConfigureMockMvc` 조합이면 전체 스프링 컨텍스트를 띄운 뒤에 MockMvc 로 컨트롤러까지 한 번에 exercising 할 수 있어서 “서비스 → 리포지토리 → 컨트롤러 흐름”을 실제 애플리케이션과 거의 같은 환경에서 검증할 수 있다는 점이 편리합니다.

다만 이번 테스트는 서비스 레이어 단독 검증만으로도 충분해 보여서, 말씀해 주신 대로 필요한 범위까지만 로드해도 될 것 같습니다.

```java
@ExtendWith(MockitoExtension.class)            // Mockito 전용 실행기
class MemberServiceTest {

    @InjectMocks
    private MemberService memberService;       // 테스트 대상

    @Mock private MemberRepository memberRepository;
    @Mock private KakaoOAuthService kakaoOAuthService;
    @Mock private TokenService tokenService;

    ...
}
```

 속도와 격리를 위해 애너테이션 범위 재조정해 보겠습니다! 😊

Gyuchool commented on src/main/java/gift/service/KakaoOAuthService.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    public KakaoUserResponseDto fetchUserInfo(String accessToken) {
+        return webClient.get()
```
만약 카카오 서버에 문제가 생겨서 응답을 5분 후에 준다면 어떻게 되나요? 고객은 5분동안 기다려야할까요?

5seonjae replied:
### “카카오 서버가 5 분 뒤에야 응답” 상황에서는 무엇이 일어나나?

1. `webClient.get()` 요청 → `block()` 대기 ( 별도 타임아웃을 주지 않았으므로 무한 대기 )

2. 5 분 동안 응답 없음 ( Reactor Netty 내부 타임아웃이 없다면 계속 기다림 )

3. 설정한 responseTimeout/connectionTimeout 가 없기 때문에 타임아웃 예외도 안 남

이로 인해 최악의 경우에는 사용자 요청 스레드가 5 분 이상 점유하면 클라이언트도 5 분 대기해야 합니다.

즉, 타임아웃을 명시적으로 걸어 두지 않으면 고객도 그대로 기다리게 됩니다.

---

### 해결 방안 – WebClient 타임아웃 지정

```java
this.webClient = builder
    .baseUrl(apiUrl)
    .clientConnector(new ReactorClientHttpConnector(
        HttpClient.create()
                  .responseTimeout(Duration.ofSeconds(3))   // 응답 타임아웃
                  .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, 2000) // 커넥션 타임아웃
    ))
    .build();
```

- `responseTimeout`  : 서버가 N 초 안에 바디를 주지 않으면 `ReadTimeoutException` → `ResourceAccessException` 래핑
- `CONNECT_TIMEOUT_MILLIS ` : TCP 연결 자체가 N ms 넘으면 `ConnectTimeoutException`

https://github.com/next-step/spring-gift-order/pull/91#discussion_r2230769321
이 피드백 반영 때 만들어 둔 **`handleTimeout(ResourceAccessException)`** 가 이 예외를 잡아 `504 Gateway Timeout` 으로 응답하므로, 클라이언트는 “카카오가 느려서 실패” 라는 사실을 3–5초 이내에 받게 됩니다.

Gyuchool replied:
잘 학습하셨군요👍 고객 경험도 안좋을 뿐더러 해당 api요청하는데 스레드를 다 소진해서 다른 api를 호출할때 문제가 발생해 서버 전체에 문제가 발생할 수 있는 위험이 생길수도 있습니다:)

Gyuchool commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -50,4 +56,23 @@ public boolean isAdmin(String email) {
         return member.getIsAdmin();
     }
+
+    @Transactional
```
`@Transactional`은 어떤 역할을 하나요?

5seonjae replied:
`@Transactional ` 은 로그인 과정 전체를 하나의 데이터베이스 트랜잭션으로 묶어 주는 스위치입니다.
`loginWithKakao()` 안에서 일어나는 두 가지 DB 작업이 정확히 같은 트랜잭션 컨텍스트에서 수행되도록 보장해 줍니다.

### `@Transactional ` 이 필요한 이유

| 문제 | `@Transactional` 있을 때 | 없을 때 |
| -- | -- | -- |
| 원자성 ( Atomicity ) [ find→save 가 끊김 없이 실행돼야 함 ] | 두 DB 연산이 한 트랜잭션으로 commit/rollback | 중간에 예외가 나면 일부만 commit 되어 중복 회원·데이터 불일치 위험 |
| 격리 수준 ( Isolation ) [ 동시 로그인 시 race condition ] | 트랜잭션 격리( 기본 : READ_COMMITTED ) 로 “동시에 INSERT 두 번” 방지 | 동일 카카오 ID 를 가진 회원이 두 번 insert 될 수 있음 |
| 지연 쓰기 ( Write‑behind ) [ JPA 1차 캐시 활용 ] | `save()` 직후 플러시되지 않아도 트랜잭션 끝에 한 번만 flush | 트랜잭션 없음 → 즉시 flush → 불필요한 IO 증가 |
| Lazy Loading | 세션이 열려 있어 지연 로딩 안전 | 세션 없이 엔티티 접근 시 `LazyInitializationException` |

`@Transactional` 이 없으면 

1. 두 사용자가 동시에 로그인·회원가입 요청을 보낼 때 `findByKakaoId` 가 둘 다 `empty` 를 반환 → `save()`가 두 번 실행돼 PK 중복 예외 발생 가능

2. 회원 저장 후 토큰 생성 과정에서 예외가 나면, 사용자는 실패했는데 DB 에는 회원이 이미 생성돼 데이터 불일치가 발생

3. 1차 캐시가 없는 상태에서 매번 즉시 DB 에 hit/flush 를 함

`@Transactional` 은 원자성·격리·일관성을 보장해 주므로, `loginWithKakao()` 처럼 “조회 후 없으면 INSERT” 같은 복합 DB 작업엔 필수입니다.

Gyuchool replied:
그렇다면 `@Transactional`이 걸려있으면 `loginWithKakao()` 메서드 시작할때부터 디비 커넥션이 물려있는 것일까요?
결국 `@Transactional`에 영향을 주는 로직은 findByKakaoId(), save() 뿐이긴 하겠군요🤔

5seonjae replied:
좋은 질문입니다! 😊
정리하면 `@Transactional` 이 걸린 시점부터 트랜잭션 컨텍스트가 만들어지지만, JDBC 커넥션은 실제로 “첫 번째 DB 접근이 일어날 때” 획득됩니다. 그리고 컨텍스트가 살아 있는 동안에는 모든 JPA 쓰기·읽기·지연 로딩·flush 작업이 동일 커넥션‧트랜잭션에 묶여 동작합니다.

> 그렇다면 `@Transactional` 이 걸려있으면 `loginWithKakao()` 메서드 시작할때부터 디비 커넥션이 물려있는 것일까요?

`@Transactional` 이 붙어 있다고 해서 메서드 진입 순간에 JDBC 커넥션까지 곧바로 획득되는 것은 아닙니다. 진짜 커넥션은 **첫 번째 DB 접근 시점** 에만 풀에서 빌려오고, 트랜잭션이 끝나면 반환됩니다.

`loginWithKakao()` 메서드의 코드 진행 과정에 대해 다시 살펴보겠습니다.

1. `loginWithKakao()` 호출 → `@Transactional` AOP 프록시 진입

    트랜잭션 시작 ( `TransactionSynchronizationManager` 에 컨텍스트 바인딩 )

2. `findByKakaoId()` 실행 시점

    JPA가 커넥션 풀에서 커넥션을 요청 → 동기화된 트랜잭션에 묶임 ( 이때 DB 커넥션이 발생합니다! )

3. 같은 메서드 내 `save()` 실행

    이미 획득한 동일 커넥션으로 SQL 수행

4. 메서드 정상 종료 또는 예외 throw ( 런타임 )

    각각 트랜잭션 commit, rollback 으로 인해, 커넥션이 반환됩니다.

Gyuchool replied:
`LazyConnectionDataSourceProxy`는 언제 사용하는건가요?🤔
제가 알기론 LazyConnectionDataSourceProxy를 사용할때만 커넥션 접근 시점에 가져오는 것으로 알고있어서요..!  확인부탁드립니다:)

Gyuchool commented on build.gradle:
```diff
@@ -23,11 +23,13 @@ dependencies {
     runtimeOnly  'io.jsonwebtoken:jjwt-jackson:0.12.6'
     runtimeOnly 'com.h2database:h2'
     testImplementation 'org.springframework.boot:spring-boot-starter-test'
+    testImplementation 'com.squareup.okhttp3:mockwebserver:4.9.3'
```
여러 mockServer들이 있을텐데 okhttp3 mockserver를 사용하신 이유가 궁금합니다!
- wireMock
- Hoverfly
- etc

5seonjae replied:
Mock HTTP 서버를 고르면서 고려했던 3가지 축은 (1) 의존성/빌드 무게, (2) WebClient와의 궁합, (3) 테스트 실행 속도였습니다.

그 기준으로 보면 OkHttp MockWebServer 가 제일 가볍고, 특히 단위 테스트 규모에 딱 맞아서 선택했습니다.

### 구체적으로 편했던 점

1. 1‑라이너 셋업

- 스프링 설정을 바꿀 필요 없이 동적 포트 를 바로 주입할 수 있었습니다.

2. 자바 코딩 스타일 스텁

- JSON 문자열 그대로 넣으면 끝이라서 WireMock DSL ( `stubFor(get(urlEqualTo(…)))` )보다 간결했습니다.

3. JVM 내 동작 → 디버깅·속도

- WireMock/Jetty 보다 기동 로그가 거의 없고, 테스트 당 2–3 ms 로 끝나서 체감 속도가 빨랐습니다.

이 테스트는 WebClient ↔ 외부 서버 호출을 단건 단위로 검증하는 용도라서, “가볍고 빠르며 코드 한파일로 끝나는” OkHttp MockWebServer가 가장 적합하다고 판단했습니다. 🚀

Gyuchool commented on README.md:
```diff
@@ -1 +1,137 @@
+JWT_EXPIRATION_MS=3600000
+
+# 3. 카카오 로그인 REST API 키 (Client ID)
+rest-api.clientId=eb0390277e09b0ef2defdaaeea2c88a8
```
요구사항에 아래와 같이 적혀있어서 번거롭겠지만 재발급하고 요청주세요!🙏
<img width="604" height="224" alt="image" src="https://github.com/user-attachments/assets/15a4e273-f79c-4641-8c27-9d6ca8ff5250" />

아시다시피 API Key같은 민감정보들은 유출되면 악용될 여지가 있기 때문에, 노출되지 않도록 주의해야합니다:)

아마도 리뷰어의 편의를 위해 코드에 남겨주신 것 같지만 DM으로 전달주신다면 감사하겠습니다🙇‍♂️
감사합니다. ☺️

5seonjae replied:
기능 요구 사항 부분을 확인했었는데, 저 부분을 놓치고 구현했던 것 같습니다.

REST API 키는 재발급 받아서 반영하고 README.md 에도 이 부분은 수정하도록 하겠습니다!

다음부터는 기능 요구 사항을 꼼꼼하게 읽어보고 구현하도록 하겠습니다.

Gyuchool reviewed (changes requested):
안녕하세요 선재님
빠르게 1단계 구현해주셨군요👍

다만, properties 파일에  `jwt.secret` 값이 없어서 그런지 실행 및 테스트가 동작 안하는데 확인 부탁드립니다!

몇 가지 코멘트를 남겨두었으니 반영 및 의견 주시고 다시 요청주세요!🙏
궁금한 점은 디스코드 DM이나 PR 커멘트로 언제든지 여쭤봐주세요!(새벽도 가능!)
끝까지 화이팅입니다💪

5seonjae committed:
chore: 코드 가독성 향상을 위해 불필요한 주석 제거

5seonjae committed:
refactor(kakao): ResponseCookie로 AUTH 쿠키 생성 로직 변경

- ResponseCookie.from("AUTH", jwt) 빌더 사용해 httpOnly, secure, path, maxAge(Duration.ofHours(1)), sameSite("Lax") 속성 설정
- 기존 `new Cookie` 방식에서 전환하여 보안 강화 및 쿠키 수명 제어 기능 추가

5seonjae committed:
refactor(auth): kakaoCallback 흐름 간소화 및 조기 반환 적용

- `code`가 null인 경우 `/login`으로 즉시 반환하도록 분기 반전
- 중첩된 `if` 블록 제거로 제어 흐름과 가독성 향상

5seonjae committed:
refactor(dto): KakaoTokenResponseDto에 @JsonNaming(SnakeCaseStrategy) 적용

- 개별 `@JsonProperty` 어노테이션 제거
- `@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)` 추가로 스네이크 케이스 네이밍 전략 일관 적용

5seonjae committed:
refactor(kakao): 카카오 OAuth 예외 처리 및 에러 매핑 로직 개선

- `KakaoOAuthException` 글로벌 핸들러 수정 및 `mapToStatus` 메서드로 HTTP 상태 매핑
- `ResourceAccessException` 발생 시 SocketTimeout 예외 감지하여 504 GATEWAY_TIMEOUT 반환
- 일반 `Exception` 핸들러 추가로 예외 일관성 확보
- `exchangeCodeForToken` 및 `fetchUserInfo`에서 `onStatus` 사용해 오류 응답 바디 포함 예외 던지도록 수정

5seonjae committed:
chore(env): .env 키 rest-api.clientId → kakao.clientId로 변경

- 환경 변수 이름 일관성 확보를 위해 `.env` 파일에서 `rest-api.clientId`를 `kakao.clientId`로 수정

5seonjae committed:
refactor(auth): loginWithKakao 메소드에서 var 타입 추론 대신 명시적 타입 선언 적용

- tokenDto, userDto, nickname 변수를 명시적 타입으로 변경하여 가독성 및 타입 안전성 강화

5seonjae committed:
test(kakao): KakaoOAuthServiceTest 및 MemberServiceTest에 @DisplayName 애노테이션 추가

- 각 테스트 메서드에 `@DisplayName` 애노테이션 적용하여 실행 결과 가독성 향상

5seonjae committed:
test(service): MemberServiceTest를 MockitoExtension 기반 단위 테스트로 전환

- `@SpringBootTest`, `@AutoConfigureMockMvc`, `@Transactional` 제거
- `@ExtendWith(MockitoExtension.class)`, `@InjectMocks`, `@Mock` 애노테이션 적용하여 경량화 및 격리성 확보

5seonjae committed:
perf(kakao): WebClient에 커스텀 타임아웃 및 커넥터 설정

- Reactor Netty 클라이언트용 `ReactorClientHttpConnector`에 CONNECT_TIMEOUT_MILLIS 2초, responseTimeout 3초 설정
- `exchangeCodeForToken` 및 `fetchUserInfo` 요청 파이프라인에 `.timeout(Duration.ofSeconds(3))` 적용하여 최대 대기 시간 제한

5seonjae committed:
docs(readme): .env 샘플 키 kakao.clientId로 변경 및 노출 주의사항 추가

- 샘플 `.env`에서 프로퍼티 키를 `rest-api.clientId`에서 `kakao.clientId`로 수정
- 프로덕션용 REST API 키는 절대 커밋하거나 공개 저장소에 노출하지 말라는 주의 문구 추가

5seonjae committed:
fix(config): OAuth 인가 URL 프로퍼티 키 rest-api.clientId → kakao.clientId로 변경

- Thymeleaf 로그인 링크에서 `client_id` 파라미터에 `@environment.getProperty('kakao.clientId')` 사용하도록 수정

Gyuchool commented on src/main/java/gift/service/KakaoOAuthService.java:
```diff
@@ -0,0 +1,93 @@
+    @Value("${kakao.api-url}")
+    private final String apiUrl;
+
+    @Value("${kakao.redirect-uri:}")
```
```suggestion
    @Value("${kakao.redirect-uri}")
```
이건 오타일까요? 디폴트 값 지정하려고 했던걸까요? 디폴트 값 지정하려면 아래처럼 할 수도 있습니다😊
```suggestion
    @Value("${kakao.redirect-uri:localhost:8080}")
```

5seonjae replied:
수정 제안 감사합니다! 😊
맞습니다 `:` 을 붙여 기본값을 줄 생각이었는데, 값까지 채워 넣는 걸 깜빡했습니다.

다시 생각해보니 redirect-uri 는 환경마다 반드시 값을 지정해야 하는 필수 설정이기 때문에 기본값 없이 두는 편이 더 안전할 것 같아서 기본값을 지정하지 않는 쪽으로 수정하도록 하겠습니다!

Gyuchool commented on src/main/java/gift/service/KakaoOAuthService.java:
```diff
@@ -0,0 +1,93 @@
+@Service
+public class KakaoOAuthService {
+
+    @Value("${kakao.clientId}")
```
```suggestion
    @Value("${kakao.client-id}")
```
여기도 일관성있게 이렇게 하면 좋겠네요!

5seonjae replied:
이 부분도 마찬가지로 일관성 있게 수정하도록 하겠습니다!

Gyuchool commented on src/main/java/gift/service/KakaoOAuthService.java:
```diff
@@ -0,0 +1,93 @@
+        this.webClient = builder
+            .clientConnector(new ReactorClientHttpConnector(
+                HttpClient.create()
+                    .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, 2_000)
+                    .responseTimeout(Duration.ofSeconds(3))
```
피드백 반영 감사합니다:)

Gyuchool commented on src/test/java/gift/KakaoOAuthServiceTest.java:
```diff
@@ -0,0 +1,109 @@
+        server.shutdown();
+    }
+
+    @Test
+    @DisplayName("[성공] 인가 코드를 액세스 토큰으로 교환한다")
```
[성공], [실패] 구분 좋네요👍

Gyuchool reviewed (approved):
안녕하세요 선재님!
1단계 미션 진행하시느라 고생하셨습니다!!
피드백을 너무 잘 반영해주시고 정리까지 너무 깔끔해서 좋았네요💯💯

몇 가지 코멘트를 남겨두었으니 다음 단계 진행하시면서 반영 및 의견 주시면 좋을 것 같습니다.🙏
궁금한 점은 디스코드 DM이나 PR 커멘트로 언제든지 여쭤봐주세요!(새벽도 가능!)
다음 미션도 화이팅입니다!! 💪
